src/common/lib/transport/connectionmanager.ts

```diff
@@ -1511,130 +1335,132 @@ class ConnectionManager extends EventEmitter {
-   * and dispatches accordingly. After a transport has been set pending,
-   * tryATransport calls connectImpl to see if there's another stage to be done.
-   * */
-  connectImpl(transportParams: TransportParams, connectCount?: number): void {
+   * handling of the case where we need to failover from web_socket to the base
+   * transport is implemented in the connectWs method.
+   */
+  connectImpl(transportParams: TransportParams, connectCount: number): void {
     const state = this.state.state;
-
-    if (state !== this.states.connecting.state && state !== this.states.connected.state) {
+    if (state !== this.states.connecting.state) {
       /* Only keep trying as long as in the 'connecting' state (or 'connected'
        * for upgrading). Any operation can put us into 'disconnected' to cancel
        * connection attempts and wait before retrying, or 'failed' to fail. */
       Logger.logAction(
         Logger.LOG_MINOR,
         'ConnectionManager.connectImpl()',
-        'Must be in connecting state to connect (or connected to upgrade), but was ' + state
-      );
-    } else if (this.pendingTransports.length) {
-      Logger.logAction(
-        Logger.LOG_MINOR,
-        'ConnectionManager.connectImpl()',
-        'Transports ' + this.pendingTransports[0].toString() + ' currently pending; taking no action'
+        'Must be in connecting state to connect, but was ' + state
       );
-    } else if (state == this.states.connected.state) {
-      this.upgradeIfNeeded(transportParams);
-    } else if (this.transports.length > 1 && this.getTransportPreference()) {
-      this.connectPreference(transportParams, connectCount);
-    } else {
-      this.connectBase(transportParams, connectCount);
+      return;
     }
-  }
 
-  connectPreference(transportParams: TransportParams, connectCount?: number): void {
-    const preference = this.getTransportPreference();
-    let preferenceTimeoutExpired = false;
+    const transportPreference = this.getTransportPreference();
 
-    if (!this.transports.includes(preference)) {
-      this.unpersistTransportPreference();
-      this.connectImpl(transportParams, connectCount);
+    // If transport preference is for a non-ws transport but websocket is now available, unpersist the preference for next time
+    if (transportPreference && transportPreference === this.baseTransport && this.webSocketTransportAvailable) {
+      this.checkWsConnectivity()
+        .then(() => {
+          this.unpersistTransportPreference();
+          if (this.state === this.states.connecting) {
+            Logger.logAction(
+              Logger.LOG_MINOR,
+              'ConnectionManager.connectImpl():',
+              'web socket connectivity available, cancelling connection attempt with ' + this.baseTransport
+            );
+            if (this.proposedTransport) {
+              this.proposedTransport.dispose();
+            }
+            this.connectWs(transportParams, this.connectCounter++);
```

Comment:
> `this.connectWs(transportParams, this.connectCounter++);`

this looks dodgy. Remember that `x++` increments x and returns the _old_ value -- I haven't had to think about the connect counter in a while but that seems unlikely to be what you intended

Reply:
oh good point, i've changed both to pre-increment now

Reply:
I haven't been involved in this PR at all so sorry for the drive-by comment, but I'd suggest that this kind of confusion about the subtle distinctions between these operators means that they're perhaps best avoided, and we could just have an extra line of code to do the increment?

Reply:
(They [were removed in Swift](https://github.com/apple/swift-evolution/blob/main/proposals/0004-remove-pre-post-inc-decrement.md) for this reason for example)